### PR TITLE
event-source: Migrate from Istio Policy to PeerAuthentication

### PR DESCRIPTION
**Pull request by nachtmaar**
**Description**

event-source: Migrate from Istio Policy to PeerAuthentication

Changes proposed in this pull request:

- migrate from API group `authentication.istio.io` to `security.istio.io`
- for the time being (Kyma 1.16) the controller does the migration part: remove `Policy.authentication.istio.io` and create a `PeerAuthentication.security.istio.io` instead
- Then Kyma 1.16 is released we can remove the code which is interacting with policies. The code which needs to be deleted after Kyma 1.16 is marked inside todo blocks to make it easier to remove it.
- `Istio Policy migrated to PeerAuthentication` is the test which tests the upgrade logic.

Further execution plan:
- As soon as this PR is reviewed, the yaml changes will be merged to <https://github.com/kyma-project/kyma/pull/8898>
- Using this PR (has istio 1.5) the tests will be checked
- If tests are greeen and the istio branch is merged, we can merge this PR to master.

**Related issue(s)**

https://github.com/kyma-project/kyma/issues/9105

### CODE REVIEW

**Comment by marcobebway** [comment]
I believe we need to update the istio version in the `Gopkg.toml` file.

File: components/event-sources/reconciler/object/peerauthentication.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+			Name:      name,
+		},
+	}
+	for _, opt := range opts {
+		opt(s)
+	}
+	return s
+}
+
+// WithPermissiveModeDeprecated sets the mTLS mode of the PeerAuthentication to Permissive for the given port
```
Comment (marcobebway) [inline components/event-sources/reconciler/object/peerauthentication.go:23]: To mark this function as deprecated please check this guide: https://github.com/golang/go/wiki/Deprecated.

File: components/event-sources/reconciler/httpsource/reconcile_test.go [hunk 2]
```diff
@@ -30,24 +34,30 @@ import (
...
+	tUID                = types.UID("00000000-0000-0000-0000-000000000000")
+	tImg                = "sources.kyma-project.io/http:latest"
+	tExternalPort       = 80
+	tPort               = 8080
+	tSinkURI            = "http://" + tName + "-kn-channel." + tNs + ".svc.cluster.local"
+	tSource             = "varkes"
+	tPeerAuthentication = "test"
+	tRevisionSvc        = "test"
+	tTargetPort         = metricsPort
+	// TODO: delete me then Kyma 1.16 is out
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/event-sources/reconciler/httpsource/reconcile_test.go:58]: I wanted to create a follow-up PR which removes everything but not merge it yet to master. But the PR can be linked in the issue then 😎

**Comment by nachtmaar** [comment]
> I believe we need to update the istio version in the `Gopkg.toml` file.

Thanks. Updated to 1.5.8.

File: components/event-sources/reconciler/object/peerauthentication.go (hunk 1 above)
Comment (nachtmaar) [reply components/event-sources/reconciler/object/peerauthentication.go]: Solved

File: components/event-sources/reconciler/httpsource/reconcile_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/event-sources/reconciler/httpsource/reconcile_test.go]: Issue created and referenced in todo comments.

File: components/event-sources/reconciler/httpsource/reconcile.go [hunk 3]
```diff
@@ -292,25 +308,45 @@ func (r *Reconciler) getOrCreateService(src *sourcesv1alpha1.HTTPSource,
 	return service, nil
 }
 
-// getOrCreatePolicy returns the existing Policy for a Revision of a KnativeService, or
+// getOrCreatePeerAuthentication returns the existing PeerAuthentication for a Revision of a Deployment, or
```
Comment (sayanh) [inline components/event-sources/reconciler/httpsource/reconcile.go:311]:
```suggestion
// getOrCreatePeerAuthentication returns the existing PeerAuthentication for a Replica of a Deployment, or
```

File: components/event-sources/reconciler/object/equality_test.go [hunk 4]
```diff
@@ -722,61 +654,46 @@ func TestHandlerEqual(t *testing.T) {
...
+				p.Spec.PortLevelMtls = map[uint32]*securityv1beta1apis.PeerAuthentication_MutualTLS{
+					differentPort: {
+						Mode: securityv1beta1apis.PeerAuthentication_MutualTLS_PERMISSIVE,
 					},
 				}
 				return p
 			},
 			expect: false,
 		},
-		"equal when labels, targets and peers are equal but other fields are different": {
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/event-sources/reconciler/object/equality_test.go:760]: True, test is back :)

**Review by sayanh — DISMISSED** [review]
lgtm

**Comment by pbochynski** [comment]
I think the code should work if deprecated API authentication.istio.io is removed. I tested this PR on istio 1.7 and it fails every second with message:
```
knative.dev/pkg/controller/controller.go:456: Failed to list *v1alpha1.Policy: the server could not find the requested resource (get policies.authentication.istio.io)
```

**Comment by nachtmaar** [comment]
> I think the code should work if deprecated API authentication.istio.io is removed. I tested this PR on istio 1.7 and it fails every second with message:
> 
> ```
> knative.dev/pkg/controller/controller.go:456: Failed to list *v1alpha1.Policy: the server could not find the requested resource (get policies.authentication.istio.io)
> ```

@pbochynski The code will get removed in the next Kyma release, see https://github.com/kyma-project/kyma/issues/9331
So far it only has to work with Istio 1.5.x

**Comment by pbochynski** [comment]
> > I think the code should work if deprecated API authentication.istio.io is removed. I tested this PR on istio 1.7 and it fails every second with message:
> > ```
> > knative.dev/pkg/controller/controller.go:456: Failed to list *v1alpha1.Policy: the server could not find the requested resource (get policies.authentication.istio.io)
> > ```
> 
> @pbochynski The code will get removed in the next Kyma release, see #9331
> So far it only has to work with Istio 1.5.x

I do not like the solution because it keeps the event-source module dependent on istio 1.5. I already use Kyma with istio 1.7 which is faster and less resource hungry. With your approach I still need my own image of event-source. Don't you think that migration job can be done in the separate binary (job) that is executed on update only? This way event source could work with any istio version >1.5 already. In the next step (kyma version) you can just delete the job.

**Review by sayanh — APPROVED** [review]

**Pull request merged**